java-checks/src/main/java/org/sonar/java/checks/xml/maven/DependencyWithSystemScopeCheck.java

```diff
@@ -19,43 +19,59 @@
  */
 package org.sonar.java.checks.xml.maven;
 
-import com.google.common.collect.ImmutableList;
-import com.google.common.collect.Lists;
-import org.apache.commons.lang.StringUtils;
+import java.util.Collections;
+import java.util.Optional;
+import javax.xml.xpath.XPathExpression;
 import org.sonar.check.Rule;
-import org.sonar.java.checks.xml.maven.helpers.MavenDependencyCollector;
-import org.sonar.java.xml.maven.PomCheck;
-import org.sonar.java.xml.maven.PomCheckContext;
-import org.sonar.maven.model.LocatedAttribute;
-import org.sonar.maven.model.maven2.Dependency;
-
-import javax.annotation.Nullable;
-import java.util.List;
+import org.sonar.java.checks.xml.AbstractXPathBasedCheck;
+import org.sonarsource.analyzer.commons.xml.XmlFile;
+import org.sonarsource.analyzer.commons.xml.XmlTextRange;
+import org.w3c.dom.Element;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
 
 @Rule(key = "S3422")
-public class DependencyWithSystemScopeCheck implements PomCheck {
+public class DependencyWithSystemScopeCheck extends AbstractXPathBasedCheck {
+
+  private XPathExpression dependencyExpression = getXPathExpression("//dependencies/dependency");
 
   @Override
-  public void scanFile(PomCheckContext context) {
-    List<Dependency> dependencies = new MavenDependencyCollector(context.getMavenProject()).allDependencies();
-    for (Dependency dependency : dependencies) {
-      LocatedAttribute scope = dependency.getScope();
-      if (scope != null && "system".equalsIgnoreCase(scope.getValue())) {
-        String message = "Update this scope.";
-        LocatedAttribute systemPath = dependency.getSystemPath();
-        List<PomCheckContext.Location> secondaries = getSecondary(systemPath);
-        if (systemPath != null) {
-          message = "Update this scope and remove the \"systemPath\".";
-        }
-        context.reportIssue(this, scope.startLocation().line(), message, secondaries);
-      }
+  protected void scanFile(XmlFile xmlFile) {
+    if (!"pom.xml".equalsIgnoreCase(xmlFile.getInputFile().filename())) {
+      return;
+    }
+
+    NodeList dependencies = evaluate(dependencyExpression, xmlFile.getNamespaceUnawareDocument());
```

Comment:
you can now rely on the new methods from `AbstracXPathBasedCheck` to directly iterate over list of nodes